### PR TITLE
shred: use clap for argument management

### PR DESCRIPTION
**Pull request by ajanicij**

### CODE REVIEW

**Comment by sylvestre** [comment]
Please run rustfmt, thanks :)

File: src/uu/shred/src/shred.rs [hunk 1]
```diff
@@ -212,136 +213,163 @@ impl<'a> BytesGenerator<'a> {
...
-        let remove = matches.opt_present("remove");
-        let size = get_size(matches.opt_str("size"));
-        let exact = matches.opt_present("exact") && size.is_none(); // if -s is given, ignore -x
-        let zero = matches.opt_present("zero");
-        let verbose = matches.opt_present("verbose");
-        for path_str in matches.free.into_iter() {
-            wipe_file(&path_str, iterations, remove, size, exact, zero, verbose);
+    }
+
+    let mut iterations: usize = 3;
```
Comment (sylvestre) [inline src/uu/shred/src/shred.rs:328]: Why you aren't using the default_values options from clap?

Comment (ajanicij) [reply src/uu/shred/src/shred.rs]: Changed, thanks! I am still new to clap, so I missed this feature.

File: src/uu/shred/src/shred.rs [hunk 2]
```diff
@@ -212,136 +213,164 @@ impl<'a> BytesGenerator<'a> {
...
-        let remove = matches.opt_present("remove");
-        let size = get_size(matches.opt_str("size"));
-        let exact = matches.opt_present("exact") && size.is_none(); // if -s is given, ignore -x
-        let zero = matches.opt_present("zero");
-        let verbose = matches.opt_present("verbose");
-        for path_str in matches.free.into_iter() {
-            wipe_file(&path_str, iterations, remove, size, exact, zero, verbose);
+    }
+
+    let mut iterations: usize = 0;
```
Comment (sylvestre) [inline src/uu/shred/src/shred.rs:329]: Please do it like:
https://github.com/uutils/coreutils/blob/4873c8a24b7e00283572f05da18e3b74a6819562/src/uu/numfmt/src/numfmt.rs#L122-L129
let iterations = match ... is more rust idiomatic

Comment (ajanicij) [reply src/uu/shred/src/shred.rs]: Thanks! I am still learning about idiomatic Rust style (and in general, I am still new to Rust).

File: src/uu/shred/src/shred.rs [hunk 3]
```diff
@@ -212,136 +213,164 @@ impl<'a> BytesGenerator<'a> {
...
-                Err(_) => {
-                    println!("{}: Invalid number of passes", NAME);
-                    return 1;
-                }
+    }
+
+    let iterations = match matches.value_of(options::ITERATIONS) {
+        Some(s) => match s.parse::<usize>() {
+            Ok(u) => u,
+            Err(_) => {
```
Comment (sylvestre) [inline src/uu/shred/src/shred.rs:332]: thanks, please run rustfmt :)

Comment (ajanicij) [reply src/uu/shred/src/shred.rs:332]: I keep forgetting it :-(

File: src/uu/shred/src/shred.rs [hunk 4]
```diff
@@ -212,136 +213,164 @@ impl<'a> BytesGenerator<'a> {
...
-        Err(e) => panic!("Invalid options\n{}", e),
+    let matches = App::new("shred")
+        .name("shred")
+        .version(VERSION_STR)
+        .about(
+            "Overwrite the specified FILE(s) repeatedly, in order to make it harder\n\
+for even very expensive hardware probing to recover the data.",
+        )
+        .after_help(AFTER_HELP)
+        .usage("shred [OPTION]... FILE...")
```
Comment (sylvestre) [inline src/uu/shred/src/shred.rs:274]: please move that into a function like 
https://github.com/uutils/coreutils/blob/master/src/uu/ls/src/ls.rs#L50-L53 
for consistency with the other binaries

File: src/uu/shred/src/shred.rs [hunk 5]
```diff
@@ -212,136 +213,164 @@ impl<'a> BytesGenerator<'a> {
...
-    opts.optflag("", "version", "output version information and exit");
-
-    let matches = match opts.parse(&args[1..]) {
-        Ok(m) => m,
-        Err(e) => panic!("Invalid options\n{}", e),
+    let matches = App::new("shred")
+        .name("shred")
+        .version(VERSION_STR)
+        .about(
+            "Overwrite the specified FILE(s) repeatedly, in order to make it harder\n\
```
Comment (sylvestre) [inline src/uu/shred/src/shred.rs:270]: please move that in a variable like https://github.com/uutils/coreutils/blob/master/src/uu/ls/src/ls.rs#L44

File: src/uu/shred/src/shred.rs [hunk 6]
```diff
@@ -212,136 +213,164 @@ impl<'a> BytesGenerator<'a> {
...
-        "add a final overwrite with zeros to hide shredding",
-    );
-    opts.optflag("", "help", "display this help and exit");
-    opts.optflag("", "version", "output version information and exit");
-
-    let matches = match opts.parse(&args[1..]) {
-        Ok(m) => m,
-        Err(e) => panic!("Invalid options\n{}", e),
+    let matches = App::new("shred")
+        .name("shred")
```
Comment (sylvestre) [inline src/uu/shred/src/shred.rs:267]:
```suggestion
    let matches = App::new(executable!())
```

File: src/uu/shred/src/shred.rs [hunk 7]
```diff
@@ -212,136 +213,164 @@ impl<'a> BytesGenerator<'a> {
...
+        .version(VERSION_STR)
+        .about(
+            "Overwrite the specified FILE(s) repeatedly, in order to make it harder\n\
+for even very expensive hardware probing to recover the data.",
+        )
+        .after_help(AFTER_HELP)
+        .usage("shred [OPTION]... FILE...")
+        .arg(Arg::with_name(options::FILE).hidden(true).multiple(true))
```
Comment (sylvestre) [inline src/uu/shred/src/shred.rs:275]: please move that at the end of the list of Arg

File: src/uu/shred/src/shred.rs [hunk 8]
```diff
@@ -212,136 +213,172 @@ impl<'a> BytesGenerator<'a> {
...
+    };
+    let size = get_size(size_arg);
+    let exact = matches.is_present(options::EXACT) && size.is_none(); // if -s is given, ignore -x
+    let zero = matches.is_present(options::ZERO);
+    let verbose = matches.is_present(options::VERBOSE);
+
+    if !errs.is_empty() {
+        eprintln!("Invalid arguments supplied.");
+        for message in errs {
+            eprintln!("{}", message);
```
Comment (sylvestre) [inline src/uu/shred/src/shred.rs:372]: can be done in a followup PR but it would be nice to use show_error()

**Pull request merged**